src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,357 @@
+        return value
+
+    def _write_registers(self, values):
+        """
+        Write to all four registers. Try to reconnect to device in case connection was lost.
+        :values (list of 4 ints): register values (-x, -y, x, y)
+        """
+        if len(values) != 4:
+            raise ValueError("write_registers received payload of invalid length %s != 4." % len(values))
+
+        # Check if values are in allowed range
+        if not all(0 <= val <= 0xFFFF for val in values):
+            raise ValueError("Register values %s not in range [0, 65535]." % values)
+
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rq = self._serial.write_registers(BEAMDEFL_LX, values, unit=SLAVE_UNIT)
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _read_registers(self):
+        """
+        Read all four registers. Try to reconnect to device in case connection was lost.
+        :return (list of 4 ints): register values (-x, -y, x, y)
+        """
+        try:
+            # write all registers together (starting at lower x register (=0x01))
+            rr = self._serial.read_holding_registers(BEAMDEFL_LX, 4, unit=SLAVE_UNIT)
+            return rr.registers
+        except IOError:
+            self._reconnect()
+            raise IOError("Failed to write registers of beam control firmware, "
+                          "restarted serial connection.")
+
+    def _reconnect(self):
+        """
+        Attempt to reconnect the camera. It will block until this happens.
+        On return, the hardware should be ready to use as before, excepted it
+        still needs the settings to be applied.
+        """
+        num_it = 5
+        self.state._set_value(model.HwError("Beam deflection controller disconnected"), force_write=True)
+        logging.warning("Failed to write registers, trying to reconnect...")
+        for i in range(num_it):
+            try:
+                self._serial.close()
+                self._serial.connect()
+                logging.info("Recovered device.")
+                break
+            except IOError:
+                continue
+        else:
+            raise IOError("Failed to reconnect to beam deflection controller.")
+        self.state._set_value(model.ST_RUNNING, force_write=True)
+
+    def updateMetadata(self, md):
+        logging.debug("Updating metadata %s." % md)
+        if model.MD_CALIB_BEAMSHIFT in md and self._metadata.get(model.MD_CALIB_BEAMSHIFT):
+            # Transform to register values with old metadata
+            xlower, ylower, xupper, yupper = self._metadata.get(model.MD_CALIB_BEAMSHIFT)
+            vals = transform_coordinates(self.shift.value, xlower, ylower, xupper, yupper)
```

Comment:
On the first time, the shift value is initialised at 0,0, but that's not the actual value. => In this case, read the values from the hardware. A simple alternative would be to always read the values from the hardware. (so no need to use the current metadata).